aquadoggo/src/graphql/schema_builders/document.rs

```diff
@@ -0,0 +1,474 @@
+        Field::new(
+            schema_id.to_string(),
+            TypeRef::named(schema_id.to_string()),
+            move |ctx| {
+                let schema_id = schema_id.clone();
+                FieldFuture::new(async move {
+                    // Parse arguments.
+                    let mut document_id = None;
+                    let mut document_view_id = None;
+                    for (name, id) in ctx.field().arguments()?.into_iter() {
+                        match name.as_str() {
+                            DOCUMENT_ID_ARG => {
+                                document_id = Some(DocumentIdScalar::from_value(id)?);
+                            }
+                            DOCUMENT_VIEW_ID_ARG => {
+                                document_view_id = Some(DocumentViewIdScalar::from_value(id)?)
+                            }
+                            _ => (),
+                        }
+                    }
+
+                    // Check a valid combination of arguments was passed.
+                    match (&document_id, &document_view_id) {
+                        (None, None) => {
+                            return Err(Error::new(
+                                "Either document id or document view id arguments must be passed",
+                            ))
+                        }
+                        (Some(_), Some(_)) => {
+                            return Err(Error::new(
+                                "Both document id and document view id arguments cannot be passed",
+                            ))
+                        }
+                        (Some(id), None) => {
+                            debug!("Query to {} received for document {}", schema_id, id);
+                        }
+                        (None, Some(id)) => {
+                            debug!(
+                                "Query to {} received for document at view id {}",
+                                schema_id, id
+                            );
+                        }
+                    };
+                    // Pass them up to the children query fields.
+                    Ok(Some(FieldValue::owned_any((document_id, document_view_id))))
+                })
+            },
+        )
+        // TODO: We'd be better off using an enum input type here i think, then we can specify it
+        // as required, which will provide a validation step for us.
+        .argument(InputValue::new(
+            DOCUMENT_ID_ARG,
+            TypeRef::named(DOCUMENT_ID_SCALAR),
+        ))
+        .argument(InputValue::new(
+            DOCUMENT_VIEW_ID_ARG,
+            TypeRef::named(DOCUMENT_VIEW_ID_SCALAR),
+        ))
+        .description(format!(
+            "Query a {} document by id or view id.",
+            schema.name()
+        )),
+    )
+}
+
+/// Add query for getting all documents of a certain schema type to the root query object.
+///
+/// Constructs an endpoint with the format `all_<SCHEMA_ID>`.
```

Comment:
Is endpoint the right word in GraphQL land? Actually can't find really what else to mention. Would it be an Operation? (https://spec.graphql.org/October2021/#sec-Language.Operations)